commons/src/test/java/io/aiven/kafka/tieredstorage/commons/storage/filesystem/FileSystemStorageTest.java

```diff
@@ -0,0 +1,181 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class FileSystemStorageTest {
+
+    @Test
+    void testRootCannotBeAFile(@TempDir final Path root) throws IOException {
+        final Path wrongRoot = root.resolve("file_instead");
+        Files.writeString(wrongRoot, "Wrong root");
+
+        assertThrows(IllegalArgumentException.class, () -> new FileSystemStorage(wrongRoot, true));
+    }
+
+    @Test
+    void testRootCannotBeNonWritableDirectory(@TempDir final Path root) throws IOException {
+        final Path nonWritableDir = root.resolve("non_writable");
+        final FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(
+            Set.of(
+                PosixFilePermission.OWNER_READ,
+                PosixFilePermission.GROUP_READ,
+                PosixFilePermission.OTHERS_READ
+            ));
+        Files.createDirectory(nonWritableDir, fileAttributes);
+
+        assertThrows(IllegalArgumentException.class, () -> new FileSystemStorage(nonWritableDir, true));
+    }
+
+    @Test
+    void testUploadANewFile(@TempDir final Path root) throws IOException {
+        final FileSystemStorage storage = new FileSystemStorage(root, false);
+        final String content = "content";
+        final String key = "key";
+        storage.upload(new ByteArrayInputStream(content.getBytes()), key);
+
+        assertEquals(content, Files.readString(root.resolve(key)));
+    }
+
+    @Test
+    void testUploadFailsWhenFileExists(@TempDir final Path root) throws IOException {
+        final String key = "key";
+        Files.writeString(root.resolve(key), "previous");
+        final FileSystemStorage storage = new FileSystemStorage(root, false);
+        final String content = "content";
+
+        assertThrows(IOException.class, () -> storage.upload(new ByteArrayInputStream(content.getBytes()), key));
+    }
+
+    @Test
+    void testUploadWithOverridesWhenFileExists(@TempDir final Path root) throws IOException {
+        final String key = "key";
+        Files.writeString(root.resolve(key), "previous");
+        final FileSystemStorage storage = new FileSystemStorage(root, true);
+        final String content = "content";
+
+        assertDoesNotThrow(() -> storage.upload(new ByteArrayInputStream(content.getBytes()), key));
+    }
+    
+    @Test
+    void testFetchAll(@TempDir final Path root) throws IOException {
+        final String key = "key";
+        final String content = "content";
+        Files.writeString(root.resolve(key), content);
+        final FileSystemStorage storage = new FileSystemStorage(root, true);
+
+        try (final InputStream fetch = storage.fetch(key)) {
+            assertArrayEquals(content.getBytes(), fetch.readAllBytes());
+        }
+    }
+
+    @Test
+    void testFetchWithOffsetRange(@TempDir final Path root) throws IOException {
+        final String key = "key";
+        final String content = "content";
```

Comment:
Same as for key, maybe constant?

Reply:
I guess this could be specific for the test case. Don't want to couple them to the content value.